mesh/mesh.go

```diff
@@ -258,77 +232,75 @@ func (msh *Mesh) setProcessedLayer(layerID types.LayerID) {
 		msh.maxProcessedLayer = layerID
 	}
 
-	if layerID != msh.processedLayer.Add(1) {
+	if layerID != processed.Add(1) {
 		msh.With().Info("trying to set processed layer out of order",
-			log.FieldNamed("processed", msh.processedLayer),
+			log.FieldNamed("processed", processed),
 			layerID)
 		msh.nextProcessedLayers[layerID] = struct{}{}
 		return
 	}
 
 	msh.nextProcessedLayers[layerID] = struct{}{}
-	lastProcessed := msh.processedLayer
 	for i := layerID; !i.After(msh.maxProcessedLayer); i = i.Add(1) {
 		_, ok := msh.nextProcessedLayers[i]
 		if !ok {
 			break
 		}
-		lastProcessed = i
+		processed = i
 		delete(msh.nextProcessedLayers, i)
 	}
-	msh.processedLayer = lastProcessed
+	msh.processedLayer.Store(processed)
 	events.ReportNodeStatusUpdate()
-	msh.Event().Info("processed layer set", msh.processedLayer)
+	msh.Event().Info("processed layer set", processed)
 
-	if err := msh.persistProcessedLayer(msh.processedLayer); err != nil {
+	if err := msh.persistProcessedLayer(processed); err != nil {
 		msh.With().Error("failed to persist processed layer",
-			log.FieldNamed("processed", lastProcessed),
+			log.FieldNamed("processed", processed),
 			log.Err(err))
 	}
 }
 
-type validator struct {
-	*Mesh
-}
-
 // ProcessLayer performs fairly heavy lifting: it triggers tortoise to process the full contents of the layer (i.e.,
 // all of its blocks), then to attempt to validate all unvalidated layers up to this layer. It also applies state for
 // newly-validated layers.
-func (vl *validator) ProcessLayer(ctx context.Context, layerID types.LayerID) error {
-	logger := vl.WithContext(ctx).WithFields(layerID)
+func (msh *Mesh) ProcessLayer(ctx context.Context, layerID types.LayerID) error {
+	msh.mu.Lock()
+	defer msh.mu.Unlock()
```

Comment:
the issue that i want to prevent is concurrent calls by sync and hare, with a different layers. whole method needs to be locked otherwise we can read one value for msh.LatestLayerInState() on L295 and the other one on L572. 

also previously we had a lock in updateStateWithLayer, so i assume we need it there as well, i just moved it up so that it is easier to reason about ProcessLayer and prevent unexpected concurrency issues